File: 0046-permutations/0046-permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    const res = [];
    function loop(i) {
        if (i === nums.length) res.push([...nums]);
        for (let j = i; j < nums.length; j++) {
            [nums[i], nums[j]] = [nums[j], nums[i]];
            loop(i + 1);
            [nums[i], nums[j]] = [nums[j], nums[i]];
        }
    }
    loop(0);
    return res;
};

File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        def loop(i):
            if i == len(nums):
                res.append(list(nums))
            for j in range(i, len(nums)):
                nums[i], nums[j] = nums[j], nums[i]
                loop(i + 1)
                nums[i], nums[j] = nums[j], nums[i]
        loop(0)
        return res